[clang] Macro for constant rounding mode

spavloff opened the pull request:
The forthcoming C standard defines pragma FENV_ROUND to support constant rounding mode. It also requires some functions to be evaluated with such mode, N3096 7.6.2p4 states:

    Within the scope of an FENV_ROUND pragma establishing a mode other
    than FE_DYNAMIC ... invocations of functions indicated in the table
    below, for which macro replacement has not been suppressed (7.1.4),
    shall be evaluated according to the specified constant rounding mode
    ... . Invocations of functions for which macro replacement has been
    suppressed and invocations of functions other than those indicated
    in the table below shall not be affected by constant rounding modes
    – they are affected by (and affect) only the dynamic mode.

The way this requirement is formulated indicates that it could be implemented using preprocessor facility. Such implementation would require a builtin macro that is set in the region where pragma FENV_ROUND is in effect and reflects constant rounding mode.

This change introduces macro __ROUNDING_MODE__, which is a string dependent on the constant rounding mode:

    FE_TOWARDZERO        "_rtz"
    FE_TONEAREST         "_rte"
    FE_DOWNWARD          "_rtp"
    FE_UPWARD            "_rtn"
    FE_TONEARESTFROMZERO "_rta"
    FE_DYNAMIC           empty string

All these values except "_rta" are OpenCL rounding mode modifiers. Default value, when no pragma FENV_ROUND is specified, is empty string. Concatenation of a function name with the builtin macro can be used to obtain name of the function variant for particular rounding mode, like "sin_rtz", or "__builtin_cos_rtd". The test "macro_rounding_mode.c" added in this change provides an example of possible use.

The macro is implemented in the same way as FLT_EVAL_METHOD, which also depends on the results of semantic analysis.

spavloff committed:
[clang] Macro for constant rounding mode

The forthcoming C standard defines pragma FENV_ROUND to support constant
rounding mode. It also requires some functions to be evaluated with such
mode, N3096 7.6.2p4 states:

    Within the scope of an FENV_ROUND pragma establishing a mode other
    than FE_DYNAMIC ... invocations of functions indicated in the table
    below, for which macro replacement has not been suppressed (7.1.4),
    shall be evaluated according to the specified constant rounding mode
    ... . Invocations of functions for which macro replacement has been
    suppressed and invocations of functions other than those indicated
    in the table below shall not be affected by constant rounding modes
    – they are affected by (and affect) only the dynamic mode.

The way this requirement is formulated indicates that it could be
implemented using preprocessor facility. Such implementation would
require a builtin macro that is set in the region where pragma
FENV_ROUND is in effect and reflects constant rounding mode.

This change introduces macro __ROUNDING_MODE__, which is a string
dependent on the constant rounding mode:

    FE_TOWARDZERO        "_rtz"
    FE_TONEAREST         "_rte"
    FE_DOWNWARD          "_rtp"
    FE_UPWARD            "_rtn"
    FE_TONEARESTFROMZERO "_rta"
    FE_DYNAMIC           empty string

All these values except "_rta" are OpenCL rounding mode modifiers.
Default value, when no pragma FENV_ROUND is specified, is empty string.
Concatenation of a function name with the builtin macro can be used to
obtain name of the function variant for particular rounding mode, like
"sin_rtz", or "__builtin_cos_rtd". The test "macro_rounding_mode.c"
added in this change provides an example of possible use.

The macro is implemented in the same way as FLT_EVAL_METHOD, which also
depends on the results of semantic analysis.

efriedma-quic commented on clang/test/Preprocessor/macro_rounding_mode.c:
```diff
@@ -0,0 +1,55 @@
+// RUN: %clang_cc1 -emit-llvm -triple i386-linux -Wno-unknown-pragmas %s -o - | FileCheck %s
```
Since this is a preprocessor testcase, can you just use -E?

spavloff replied:
No, this macro is managed by the code in Sema, because pragma FENV_ROUND is processed there. If only `-E` is specified, __ROUNDING_MODE__ is always an empty string. We handle __FLT_EVAL_METHOD__ similarly.

efriedma-quic replied:
Is there some reason the preprocessor can't parse FENV_ROUND?  Breaking code with -save-temps etc. seems bad.

spavloff replied:
Parsing pragma FENV_ROUND has been moved to the preprocessor.

spavloff committed:
Move handling pragma FENV_ROUND to Preprocessor

As the macro __ROUNDING_MODE__ depends on the current static rounding
mode, which is managed by pragma FENV_ROUND, handling this pragma in
Parser is not possible anymore. Running clang with the option -E should
produce usable source code, where __ROUNDIND_MODE__ is substituted and
Parser is not called in this case. So the handler of pragma FENV_ROUND
is moved from Parser to Preprocessor.

Maintaining the rounding mode in Preprocessor also requires to keep
track of curly brace nesting level, because the pragma can be specified
inside a block, in this case its effect ends as the block is finished
and the previous static rounding mode must be established.

efriedma-quic reviewed (commented):
This approach seems much better.

spavloff committed:
Count braces only if pragma FENV_ROUND acts

efriedma-quic reviewed (approved):
LGTM, but maybe wait a few days to merge in case someone else has comments.

spavloff commented:
Thanks! I will commit it in a couple of day, if no additional feedback is provided.

efriedma-quic commented:
Oh, I somehow thought the macro was part of the spec; reading again, I guess it isn't, it's just an attempt to implement the spec.

We probably want some feedback from libc implementers to check if this is what they want.  I don't really want to end up in a situation where we end up implementing this, but then nobody ends up using it.  Or we end up with two almost identical macros for the same thing.

spavloff committed:
Use identifiers for __ROUNDING_MODE__

spavloff commented:
The discussion for the proposed mechanism is open here: https://discourse.llvm.org/t/rfc-calling-functions-if-pragma-fenv-round-is-present/79372

> > This change introduces macro ROUNDING_MODE, which is a string dependent on the constant rounding mode
>
> It expands to an identify, not a string literal, right?

Exactly.
